<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genomics • SIMplyBee</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genomics">
<meta property="og:description" content="SIMplyBee">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMplyBee</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/A_Introduction_to_honeybee_biology.html">Introduction to honeybee biology</a>
    </li>
    <li>
      <a href="../articles/C_Colony_events.html">Colony Events</a>
    </li>
    <li>
      <a href="../articles/B_Multiple_colonies.html">Multiple colonies</a>
    </li>
    <li>
      <a href="../articles/D_Crossing.html">Crossing</a>
    </li>
    <li>
      <a href="../articles/E_Genomics.html">Genomics</a>
    </li>
    <li>
      <a href="../articles/G_Sampling_functions.html">Sampling functions</a>
    </li>
    <li>
      <a href="../articles/F_Quantitative_Genetics.html">Quantitative Genetics</a>
    </li>
    <li>
      <a href="../articles/Z_FAQs.html">FAQs</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HighlanderLab/SIMplyBee/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genomics</h1>
            
            <h4 data-toc-skip class="date">2023-12-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HighlanderLab/SIMplyBee/blob/HEAD/vignettes/E_Genomics.Rmd" class="external-link"><code>vignettes/E_Genomics.Rmd</code></a></small>
      <div class="hidden name"><code>E_Genomics.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates how SIMplyBee manages and manipulates the
honey bee’s genomic information. Specifically, it describes:</p>
<ul>
<li>how to obtain the genomic information,</li>
<li>how to pool genotypes, and</li>
<li>how to compute genomic relationship matrices.</li>
</ul>
<p>Let’s first create a colony.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"SIMplyBee"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: AlphaSimR</span></span>
<span><span class="co">#&gt; Loading required package: R6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SIMplyBee'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     split</span></span>
<span><span class="va">founderGenomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://gaynorr.github.io/AlphaSimR/reference/quickHaplo.html" class="external-link">quickHaplo</a></span><span class="op">(</span>nInd <span class="op">=</span> <span class="fl">2</span>, nChr <span class="op">=</span> <span class="fl">3</span>, segSites <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">SP</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SimParamBee.html">SimParamBee</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">founderGenomes</span><span class="op">)</span></span>
<span><span class="va">SP</span><span class="op">$</span><span class="fu">setTrackRec</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># request recombination tracking</span></span>
<span></span>
<span><span class="va">baseQueens</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createCastePop.html">createVirginQueens</a></span><span class="op">(</span><span class="va">founderGenomes</span><span class="op">)</span></span>
<span><span class="va">baseDrones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createCastePop.html">createDrones</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">baseQueens</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, nInd <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="va">colony</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createColony.html">createColony</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">baseQueens</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">colony</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cross.html">cross</a></span><span class="op">(</span><span class="va">colony</span>, drones <span class="op">=</span> <span class="va">baseDrones</span><span class="op">)</span></span>
<span><span class="va">colony</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildUp.html">buildUp</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="obtaining-genomic-information">Obtaining genomic information<a class="anchor" aria-label="anchor" href="#obtaining-genomic-information"></a>
</h2>
<p>Honeybees have a haplo-diploid inheritance system where queens and
workers are diploid and drones are haploid. In SIMplyBee, we simulate
drones as doubled-haploids, that is, as fully homozygous diploid
individuals. This means that they have two identical sets of
chromosomes. When they produce sperm, their gametes all have the same
one set of chromosomes. Despite them being diploid, we generally return
a haploid set of chromosomes from drones, unless specifically requested
that you want the doubled-haploid genotype.</p>
<p>Following AlphaSimR, SIMplybee has a group of genome retrieval
functions <code>get*Haplo/Geno()</code> which extract haplotypes and
genotypes for all segregating sites (<code>SegSites</code>),
quantitative trait loci (<code>QTL</code>), markers (<code>SNP</code>),
and the identical by descent (<code>IBD</code>) haplotypes. Here, site,
locus and marker are all synonyms for a position in the genome. These
functions leverage AlphaSimR functionality, but work with SIMplyBee’s
<code>Colony</code> or <code>MultiColony</code> objects and in addition
take the <code>caste</code> argument to extract information for a
specific caste. Another argument you can use with this function is
<code>collapse = TRUE/FALSE</code>. If <code>collapse = TRUE</code> then
all of the information is collapsed together and a single matrix is
returned, if <code>collapse = FALSE</code> we return a list by caste or
by colony.</p>
<p>We recommend that you study the index of available
<code>get*()</code> functions in SIMplyBee and read this vignette for a
short demonstration: <code><a href="../reference/SIMplyBee-package.html">help(SIMplyBee)</a></code>.</p>
<p>To show all this functionality, let’s get haplotypes and genotypes
across the segregating sites for the different castes using
<code>getSegSitesGeno()</code> or <code>getSegSitesHaplo()</code>. The
first row of the output shows marker identifications (chromosome_locus)
and the first column shows haplotype identifications
(individual_haplotype). The alleles are represented with a sequence of
0’s and 1’s. Let’s first obtain the information at the segregating sites
for the queen (we limit the output to the first 10 sites):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"queen"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;     1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 2_1   1   0   1   1   1   0   0   1   0    1</span></span>
<span><span class="co">#&gt; 2_2   1   1   1   0   1   0   1   1   0    0</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"queen"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  1_1  1_2  1_3  1_4  1_5  1_6  1_7  1_8  1_9 1_10 </span></span>
<span><span class="co">#&gt;    2    1    2    1    2    0    1    2    0    1</span></span></code></pre></div>
<p>Now for the fathers:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"fathers"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;      1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 4_1    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 10_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 16_1   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 5_1    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 11_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 7_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 13_1   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 3_1    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 9_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 8_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 12_1   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 15_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 14_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 6_1    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 17_1   1   1   1   0   0   1   0   1   0    1</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"fathers"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;    1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 6    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 11   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 7    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 12   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 16   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 17   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 14   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 8    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 10   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 13   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 9    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 5    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 3    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 4    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 15   1   1   1   0   0   1   0   1   0    1</span></span></code></pre></div>
<p>Since father are drones, and these are haploid, we get one row per
father. We can retrieve the doublet-haploid (diploid implementation)
state, if this is desired (showing just one father to show this
clearly):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"fathers"</span>, </span>
<span>                nInd <span class="op">=</span> <span class="fl">1</span>, dronesHaploid <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  1_1  1_2  1_3  1_4  1_5  1_6  1_7  1_8  1_9 1_10 </span></span>
<span><span class="co">#&gt;    1    1    1    0    0    1    0    1    0    1</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"fathers"</span>, </span>
<span>               nInd <span class="op">=</span> <span class="fl">1</span>, dronesHaploid <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="co">#&gt;    1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 14   2   2   2   0   0   2   0   2   0    2</span></span></code></pre></div>
<p>Now two workers:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"workers"</span>, nInd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;      1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 24_1   1   1   1   0   1   0   1   1   0    1</span></span>
<span><span class="co">#&gt; 24_2   1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 20_1   1   0   1   1   1   0   0   1   0    1</span></span>
<span><span class="co">#&gt; 20_2   1   1   1   0   0   1   1   0   1    0</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"workers"</span>, nInd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;    1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 37   2   1   2   1   1   1   1   1   1    1</span></span>
<span><span class="co">#&gt; 42   2   2   2   0   1   1   2   1   1    0</span></span></code></pre></div>
<p>And finally four drones:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"drones"</span>, nInd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;       1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 176_1   1   1   1   0   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 189_1   1   0   1   1   1   0   0   1   0    1</span></span>
<span><span class="co">#&gt; 148_1   1   0   1   1   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 125_1   1   0   1   1   1   0   0   1   0    1</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"drones"</span>, nInd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;     1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 212   1   1   1   0   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 149   1   1   1   0   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 153   1   1   1   0   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 152   1   1   1   0   1   0   1   1   0    0</span></span></code></pre></div>
<p>You can also use <code>caste = "all"</code> to get the haplotypes and
phenotypes from every individual in the colony. If the argument
<code>collapse</code> is set to <code>FALSE</code>, then the function
returns a list with haplotypes for each caste. Let’s explore the
structure of the output:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"all"</span>, collapse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ queen       : int [1:2, 1:300] 1 1 0 1 1 1 1 0 1 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "2_1" "2_2"</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:300] "1_1" "1_2" "1_3" "1_4" ...</span></span>
<span><span class="co">#&gt;  $ fathers     : int [1:15, 1:300] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:15] "5_1" "12_1" "15_1" "10_1" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:300] "1_1" "1_2" "1_3" "1_4" ...</span></span>
<span><span class="co">#&gt;  $ workers     : int [1:200, 1:300] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:200] "34_1" "34_2" "50_1" "50_2" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:300] "1_1" "1_2" "1_3" "1_4" ...</span></span>
<span><span class="co">#&gt;  $ drones      : int [1:100, 1:300] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:100] "148_1" "169_1" "190_1" "150_1" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:300] "1_1" "1_2" "1_3" "1_4" ...</span></span>
<span><span class="co">#&gt;  $ virginQueens: NULL</span></span></code></pre></div>
<p>If the argument <code>collapse</code> is set to <code>TRUE</code>,
the function returns a single matrix with haplotypes of all the
individuals. The same behaviour is implemented for all the functions
that extract genomic information</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"all"</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  int [1:317, 1:300] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:317] "2_1" "2_2" "6_1" "12_1" ...</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:300] "1_1" "1_2" "1_3" "1_4" ...</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteHaplo.html">getSegSiteHaplo</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"all"</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;      1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 2_1    1   0   1   1   1   0   0   1   0    1</span></span>
<span><span class="co">#&gt; 2_2    1   1   1   0   1   0   1   1   0    0</span></span>
<span><span class="co">#&gt; 9_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 17_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 5_1    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 10_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 14_1   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 7_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 8_1    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 12_1   1   1   1   0   0   1   1   0   1    0</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"all"</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;    1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 2    2   1   2   1   2   0   1   2   0    1</span></span>
<span><span class="co">#&gt; 10   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 5    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 7    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 14   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 17   1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 4    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 3    1   1   1   0   0   1   0   1   0    1</span></span>
<span><span class="co">#&gt; 8    1   1   1   0   0   1   1   0   1    0</span></span>
<span><span class="co">#&gt; 11   1   1   1   0   0   1   0   1   0    1</span></span></code></pre></div>
<p>SIMplyBee also has shortcuts for these haplotype and genotype
functions to make life a bit easier for the user:</p>
<ul>
<li><p><code>getQueenSegSitesHaplo()</code></p></li>
<li><p><code>getQueenSegSitesGeno()</code></p></li>
<li><p><code>getFathersSegSitesHaplo()</code></p></li>
<li><p><code>getFathersSegSitesGeno()</code></p></li>
<li><p><code>getWorkersSegSitesHaplo()</code></p></li>
<li><p><code>getWorkersSegSitesGeno()</code></p></li>
<li><p><code>getDronesSegSitesHaplo()</code></p></li>
<li><p><code>getDronesSegSitesGeno()</code></p></li>
<li><p><code>getVirginQueensSegSitesHaplo()</code></p></li>
<li><p><code>getVriginQueensSegSitesGeno()</code></p></li>
</ul>
<p>Similar aliases exist also for extracting information about
quantitative trait loci (<code>QTL</code>), markers (<code>SNP</code>),
and the identical by descent (<code>IBD</code>) haplotypes.</p>
</div>
<div class="section level2">
<h2 id="pooling-genotypic-information">Pooling genotypic information<a class="anchor" aria-label="anchor" href="#pooling-genotypic-information"></a>
</h2>
<p>Unfortunately, in real life it’s challenging to get the genotype of
every individual honeybee and so SIMplyBee provides the function
<code><a href="../reference/getPooledGeno.html">getPooledGeno()</a></code> to imitate real life data.
<code><a href="../reference/getPooledGeno.html">getPooledGeno()</a></code> returns a pooled genotype from individual
genotypes to mimic the genotyping of a pool of colony members. A
comparison of pooled and individual genotypes also allows the user to
compare the two and see the impact of pooled samples on results.</p>
<p>Firstly let’s obtain the genotypes of the workers and of the queen so
that they’re easier to work with:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genoQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"queen"</span><span class="op">)</span></span>
<span><span class="va">genoW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"workers"</span><span class="op">)</span></span></code></pre></div>
<p>The function <code><a href="../reference/getPooledGeno.html">getPooledGeno()</a></code> required also the sex of
individuals whose genotype are getting pooled (<code>F</code> for
females and <code>M</code> for males).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sexW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCasteSex.html">getCasteSex</a></span><span class="op">(</span><span class="va">colony</span>, caste <span class="op">=</span> <span class="st">"workers"</span><span class="op">)</span></span></code></pre></div>
<p>You have two options when choosing what kind of pooled genotypes you
would like, using the <code>type =</code> argument. You can use
<code>type = "mean"</code> for the average genotypes and
<code>type = "count"</code> for the counts of reference and alternative
alleles.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getPooledGeno.html">getPooledGeno</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">genoW</span>, type <span class="op">=</span> <span class="st">"count"</span>, sex <span class="op">=</span> <span class="va">sexW</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;   1_1 1_2 1_3 1_4 1_5 1_6 1_7 1_8 1_9 1_10</span></span>
<span><span class="co">#&gt; 0   0  52   0 148 100 100 109  42 158   88</span></span>
<span><span class="co">#&gt; 1 200 148 200  52 100 100  91 158  42  112</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">poolW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPooledGeno.html">getPooledGeno</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">genoW</span>, type <span class="op">=</span> <span class="st">"mean"</span>, sex <span class="op">=</span> <span class="va">sexW</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  1_1  1_2  1_3  1_4  1_5  1_6  1_7  1_8  1_9 1_10 </span></span>
<span><span class="co">#&gt; 2.00 1.48 2.00 0.52 1.00 1.00 0.91 1.58 0.42 1.12</span></span></code></pre></div>
<p>Now lets plot and compare the pooled workers to the queen’s genotype
(note the use of jitter for queen’s genotype on the x-axis so we can
spread out the dots in the plot!).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">poolW</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="va">genoQ</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Average allele dosage in workers"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Allele dosage in the queen"</span> <span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/plot%20genoQ%20with%20poolW-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="computing-genomic-relationship-matrices">Computing Genomic Relationship Matrices<a class="anchor" aria-label="anchor" href="#computing-genomic-relationship-matrices"></a>
</h2>
<p>This section introduces the calculations of IBD and IBS genomic
relationship matrices, so let’s have a quick reminder of what these
mean. Identity-by-state (IBS) is a term used when two alleles, two
segments or sequences of the genome are identical. Identity-by-descent
(IBD) is when a segment of matching (IBS) DNA shared by two or more
individuals has been inherited from a common ancestor.</p>
<p>Using IBD and IBS can allow a user to look into the relationships
based on the genomic data. We’ll demonstrate this by calculating some
Genomic Relationship Matrices (GRM) using SIMplyBee’s
<code><a href="../reference/calcBeeGRMIbs.html">calcBeeGRMIbs()</a></code> and <code><a href="../reference/calcBeeGRMIbd.html">calcBeeGRMIbd()</a></code>.</p>
<p>Let’s look at the <code><a href="../reference/calcBeeGRMIbs.html">calcBeeGRMIbs()</a></code> first. This function
returns a Genomic Relatedness Matrix (GRM) for honeybees from IBS
genomic data (bi-allelic SNP represented as allele dosages) following
the method for the sex X chromosome (Druet and Legarra, 2020).</p>
<p>To see this, let’s obtain the genotypes and sex information of all
individuals in the colony.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSegSiteGeno.html">getSegSiteGeno</a></span><span class="op">(</span><span class="va">colony</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCasteSex.html">getCasteSex</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">colony</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s calculate the IBS GRM, we will use the genotypes to
calculate this:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GRM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBeeGRMIbs.html">calcBeeGRMIbs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">geno</span>, sex <span class="op">=</span> <span class="va">sex</span><span class="op">)</span></span></code></pre></div>
<p>This produces a matrix that we can plot and summarise - its useful to
summarise diagonal and off-diagonal values separately.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://Matrix.R-forge.R-project.org" class="external-link">"Matrix"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">GRM</span>, <span class="st">"Matrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20diagonal-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">GRM</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20diagonal-2.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;  0.4083  0.4469  0.5701  0.5645  0.6792  0.7877</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">GRM</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">lower.tri</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GRM</span>, diag <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20non-diagonal-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">#&gt; -0.431843 -0.124737 -0.009341 -0.002625  0.105393  0.644691</span></span></code></pre></div>
<p>We can also inspect GRM elements between specific caste members:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCasteId.html">getCasteId</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span> </span>
<span><span class="va">idQueen</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">queen</span></span>
<span><span class="va">idFathers</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">fathers</span></span>
<span><span class="va">idWorkers</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">workers</span></span>
<span><span class="va">idDrones</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">drones</span></span>
<span><span class="va">idVirginQueens</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">virginQueens</span></span>
<span><span class="va">mw</span> <span class="op">&lt;-</span> <span class="st">"mw"</span></span>
<span><span class="va">md</span> <span class="op">&lt;-</span> <span class="st">"md"</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">GRM</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRM</span><span class="op">[</span><span class="va">idQueen</span>, <span class="va">idFathers</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20fathers%201-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRM</span><span class="op">[</span><span class="va">idQueen</span>, <span class="va">idWorkers</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20workers%201-1.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRM</span><span class="op">[</span><span class="va">idQueen</span>, <span class="va">idDrones</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20drones%201-1.png" width="700"></p>
<p><code><a href="../reference/calcBeeGRMIbs.html">calcBeeGRMIbs()</a></code> uses the
<code><a href="../reference/calcBeeGRMIbs.html">calcBeeAlleleFreq()</a></code> function to calculate allele
frequencies for centering the honeybee genotypes. You can also use it in
some other cases:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">alleleFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBeeGRMIbs.html">calcBeeAlleleFreq</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">geno</span>, sex <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/alleleFreq-1.png" width="700"></p>
<p>Now lets look at <code><a href="../reference/calcBeeGRMIbd.html">calcBeeGRMIbd()</a></code>. This function creates
Genomic Relatedness Matrix (GRM) for honeybees based on
Identical-By-Descent (IBD) information. It returns a list with a matrix
of gametic relatedness coefficients (between genomes) and a matrix of
individual relatedness coefficients (between individuals). Please refer
to Grossman and Eisen (1989), Fernando and Grossman (1989), Fernando and
Grossman (1990), Van Arendonk, Tier, and Kinghorn (1994), and Hill and
Weir (2011) for the background on this function.</p>
<p>Now obtain the IBD haplotypes and compute IBD GRM.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haploQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIbdHaplo.html">getQueenIbdHaplo</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span>
<span><span class="va">haploF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIbdHaplo.html">getFathersIbdHaplo</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span>
<span><span class="va">haploW</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIbdHaplo.html">getWorkersIbdHaplo</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span>
<span><span class="va">haploD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIbdHaplo.html">getDronesIbdHaplo</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span>
<span><span class="va">haploV</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getIbdHaplo.html">getVirginQueensIbdHaplo</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">haplo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">haploQ</span>, <span class="va">haploF</span>, <span class="va">haploW</span>, <span class="va">haploD</span>, <span class="va">haploV</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GRMs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcBeeGRMIbd.html">calcBeeGRMIbd</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">haplo</span><span class="op">)</span></span></code></pre></div>
<p>Let’s view this matrix:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">genome</span>, <span class="st">"Matrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20calcGRMIbd-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span>, <span class="st">"Matrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20calcGRMIbd-2.png" width="700"></p>
<p>Now we can look at the diagonal of the obtained matrices that
represent 1 for a genomes and 1 + inbreeding coefficient
individuals.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">genome</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">#&gt; -0.431843 -0.124737 -0.009341 -0.002625  0.105393  0.644691</span></span>
<span> </span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;  0.5000  0.5000  0.5000  0.7338  1.0000  1.0000</span></span></code></pre></div>
<p>And now the non-diagonals that represent the coefficients of
relationship between genomes or between individuals.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">GRMs</span><span class="op">$</span><span class="va">genome</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">lower.tri</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GRMs</span><span class="op">$</span><span class="va">genome</span>, diag <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20non-diagonal1-1.png" width="700"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;  0.0000  0.0000  0.2633  0.2720  0.5200  1.0000</span></span>
<span>  </span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">lower.tri</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span>, diag <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/view%20non-diagonal1-2.png" width="700"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;  0.0000  0.1983  0.2733  0.2934  0.3700  0.9700</span></span></code></pre></div>
<p>Let’s now compare compare relationships between caste members within
a colony.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtains caste member IDs</span></span>
<span><span class="va">qI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCastePop.html">getQueen</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span><span class="op">@</span><span class="va">id</span></span>
<span><span class="va">fI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCastePop.html">getFathers</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span><span class="op">@</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">wI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCastePop.html">getWorkers</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span><span class="op">@</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">dI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCastePop.html">getDrones</a></span><span class="op">(</span><span class="va">colony</span><span class="op">)</span><span class="op">@</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">)</span></span></code></pre></div>
<p>Compare queen and fathers:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">[</span><span class="va">fI</span>, <span class="va">qI</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20fathers-1.png" width="700"></p>
<p>Queen and workers:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">[</span><span class="va">wI</span>, <span class="va">qI</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20workers-1.png" width="700"></p>
<p>Queen and drones:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">GRMs</span><span class="op">$</span><span class="va">indiv</span><span class="op">[</span><span class="va">dI</span>, <span class="va">qI</span><span class="op">]</span>, xlim <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="E_Genomics_files/figure-html/Queen%20vs%20drones-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Druet and Legarra (2020) Theoretical and empirical comparisons of
expected and realized relationships for the X-chromosome. Genetics
Selection Evolution, 52:50. <a href="https://doi.org/10.1186/s12711-020-00570-6" class="external-link uri">https://doi.org/10.1186/s12711-020-00570-6</a></p>
<p>Grossman and Eisen (1989) Inbreeding, coancestry, and covariance
between relatives for X-chromosomal loci. The Journal of Heredity,
80(2):137–142. <a href="https://doi.org/10.1093/oxfordjournals.jhered.a110812" class="external-link uri">https://doi.org/10.1093/oxfordjournals.jhered.a110812</a></p>
<p>Fernando and Grossman (1989) Covariance between relatives for
X-chromosomal loci in a population in disequilibrium. Theoretical and
Applied Genetics, 77:311–319. <a href="https://doi.org/10.1007/bf00305821" class="external-link uri">https://doi.org/10.1007/bf00305821</a></p>
<p>Fernando and Grossman (1990) Genetic evaluation with autosomal and
X-chromosomal inheritance. Theoretical and Applied Genetics, 80:75–80.
<a href="https://doi.org/10.1007/bf00224018" class="external-link uri">https://doi.org/10.1007/bf00224018</a></p>
<p>Van Arendonk, Tier, and Kinghorn (1994) Use of multiple genetic
markers in prediction of breeding values. Genetics, 137(1):319–329. <a href="https://doi.org/10.1093/genetics/137.1.319" class="external-link uri">https://doi.org/10.1093/genetics/137.1.319</a></p>
<p>Hill and Weir (2011) Variation in actual relationship as a
consequence of Mendelian sampling and linkage. Genetics Research,
93(1):47–64. <a href="https://doi.org/10.1017/s0016672310000480" class="external-link uri">https://doi.org/10.1017/s0016672310000480</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jana Obšteter, Laura Strachan, Jernej Bubnič, Gregor Gorjanc.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
